internal/client/db/backend.go

```diff
@@ -0,0 +1,1834 @@
+	"github.com/ChainSafe/gossamer/internal/client/db/metakeys"
+	"github.com/ChainSafe/gossamer/internal/client/db/offchain"
+	statedb "github.com/ChainSafe/gossamer/internal/client/state-db"
+	hashdb "github.com/ChainSafe/gossamer/internal/hash-db"
+	"github.com/ChainSafe/gossamer/internal/log"
+	memorydb "github.com/ChainSafe/gossamer/internal/memory-db"
+	"github.com/ChainSafe/gossamer/internal/primitives/blockchain"
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	p_offchain "github.com/ChainSafe/gossamer/internal/primitives/core/offchain"
+	"github.com/ChainSafe/gossamer/internal/primitives/database"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime/generic"
+	statemachine "github.com/ChainSafe/gossamer/internal/primitives/state-machine"
+	"github.com/ChainSafe/gossamer/internal/primitives/storage"
+	"github.com/ChainSafe/gossamer/internal/primitives/storage/keys"
+	"github.com/ChainSafe/gossamer/internal/primitives/trie"
+	"github.com/ChainSafe/gossamer/internal/primitives/trie/cache"
+	"github.com/ChainSafe/gossamer/internal/saturating"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"github.com/ChainSafe/gossamer/pkg/trie/triedb"
+	"golang.org/x/exp/slices"
+)
+
+var logger = log.NewFromGlobal(log.AddContext("client", "db"))
+
+// Block pruning settings.
+type BlocksPruning interface {
+	isBlocksPruning()
+}
+type BlocksPruningValues interface {
+	BlocksPruningKeepAll | BlocksPruningKeepFinalized | BlocksPruningSome
+}
+
+// Keep full block history, of every block that was ever imported.
+type BlocksPruningKeepAll struct{}
+
+// Keep full finalized block history.
+type BlocksPruningKeepFinalized struct{}
+
+// Keep N recent finalized blocks.
+type BlocksPruningSome uint32
+
+func (BlocksPruningKeepAll) isBlocksPruning()       {}
+func (BlocksPruningKeepFinalized) isBlocksPruning() {}
+func (BlocksPruningSome) isBlocksPruning()          {}
+
+// Where to find the database..
+// NOTE: only uses a custom already-open database.
+type DatabaseSource struct {
+	// the handle to the custom storage
+	DB database.Database[hash.H256]
+	// if set, the create flag will be required to open such datasource
+	RequireCreateFlag bool
+}
+
+// DB-backed patricia trie state, transaction type is an overlay of changes to commit.
+type DBState[H runtime.Hash, Hasher runtime.Hasher[H]] struct {
+	*statemachine.TrieBackend[H, Hasher]
+}
+
+// A reference tracking state.
+//
+// It makes sure that the hash we are using stays pinned in storage
+// until this structure is dropped.
+type refTrackingState[H runtime.Hash, Hasher runtime.Hasher[H]] struct {
+	state      DBState[H, Hasher]
+	storage    storageDB[H]
+	parentHash *H // can be nil
+}
+
+func (rts *refTrackingState[H, Hasher]) Drop() {
+	if rts.parentHash != nil {
+		rts.storage.StateDB.Unpin(*rts.parentHash)
+	}
+}
+
+// Database settings.
+type DatabaseSettings struct {
```

Comment:
just for consistency consider
```suggestion
type DatabaseConfig struct {
```